inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -4103,48 +4090,102 @@ public <T> BeanRegistration<T> getInFlightBean(BeanIdentifier beanIdentifier) {
+            if (reference == null) {
+                return false;
+            }
+            if (referenceEnabled == null) {
+                if (reference.isEnabled(context, resolutionContext)) {
+                    referenceEnabled = true;
+                } else {
+                    referenceEnabled = false;
+                    reference = null;
+                }
             }
-            return hierarchy;
+            return referenceEnabled;
         }
 
-        @Override
-        protected void onNewSegment(Segment<?, ?> segment) {
-            Argument<?> argument = segment.getArgument();
-            if (argument.isContainerType()) {
-                argument = argument.getFirstTypeVariable().orElse(null);
-                if (argument == null) {
-                    return;
+        public boolean isDisabled() {
+            return reference == null || referenceEnabled != null && !referenceEnabled || definitionEnabled != null && !definitionEnabled;
+        }
+
+        public boolean isDefinitionEnabled(DefaultBeanContext defaultBeanContext) {
+            return isDefinitionEnabled(defaultBeanContext, null);
+        }
+
+        public boolean isDefinitionEnabled(DefaultBeanContext context, @Nullable BeanResolutionContext resolutionContext) {
+            if (definitionEnabled == null) {
+                if (isReferenceEnabled(context, resolutionContext)) {
+                    definition = getDefinition(context);
+                    if (definition.isEnabled(context, resolutionContext)) {
+                        definitionEnabled = true;
+                    } else {
+                        definitionEnabled = false;
+                        definition = null;
+                    }
+                } else {
+                    definitionEnabled = false;
                 }
             }
-            if (argument.isProvider()) {
-                return;
+            return definitionEnabled;
+        }
+
+        public <T> BeanDefinitionReference<T> getReference() {
+            if (reference == null || referenceEnabled == null || !referenceEnabled) {
+                throw new IllegalStateException("The reference is not enabled");
             }
-            for (Map.Entry<BeanDefinition<?>, Argument<?>> entry : beanCreationTargets.entrySet()) {
-                if (argument.isAssignableFrom(entry.getValue())) {
-                    foundTargets.computeIfAbsent(entry.getKey(), bd -> new ArrayList<>(5))
-                            .add(getHierarchy());
+            return reference;
+        }
+
+        public <T> BeanDefinition<T> getDefinition(BeanContext beanContext) {
+            if (definitionEnabled != null && !definitionEnabled) {
+                throw new IllegalStateException("The definition is not enabled");
+            }
+            try {
+                if (definition == null) {
```

Comment:
this is racey with isDefinitionEnabled

Reply:
We don't really care, it can be loaded multiple times like it's now